sch_yang_library.c

```diff
@@ -0,0 +1,475 @@
+
+    g_free (state_str);
+    return state;
+}
+
+/**
+ * Remove an entry from the apteryx database for the specified model
+ *
+ * @param loaded - Structure with the relevant model information
+ */
+void
+yang_library_remove_model_information (sch_loaded_model *loaded)
+{
+    char *path = g_strdup_printf ("%s/%s", YANG_LIBRARY_MOD_SET_COMMON_MOD, loaded->model);
+    apteryx_prune (path);
+    g_free (path);
+}
+
+/**
+ * Update the features leaf-list using the updated models features
+ *
+ * @param loaded A control structure for a loaded model
+ */
+void
+yang_library_update_feature_information (sch_loaded_model *loaded)
+{
+    GNode *root;
+    GNode *modules;
+    GNode *gnode;
+
+    char *path = g_strdup_printf ("%s/%s/feature", YANG_LIBRARY_MOD_SET_COMMON_MOD, loaded->model);
+    apteryx_prune (path);
+    g_free (path);
+
+    root = APTERYX_NODE (NULL, g_strdup (YANG_LIBRARY_PATH));
+    modules = add_leaf_strdup (root, YANG_LIBRARY_SCHEMA_MODULE_SET, COMMON_STR);
+    add_leaf_strdup (modules, YANG_LIBRARY_MODULE_SET_NAME, COMMON_STR);
+    gnode = add_leaf_strdup (modules, YANG_LIBRARY_MODULE_SET_MODULE_PATH,
+                             loaded->model);
+    if (loaded->features)
+    {
+        gchar **split;
+        int count;
+        int i;
+
+        split = g_strsplit (loaded->features, ",", 0);
+        count = g_strv_length (split);
+        for (i = 0; i < count; i++)
+        {
+            char *feature_path;
+            feature_path = g_strdup_printf ("%s/%s",
+                                            YANG_LIBRARY_MODULE_SET_MODULE_FEATURE,
+                                            split[i]);
+            add_leaf_strdup (gnode, feature_path, split[i]);
+            g_free (feature_path);
+        }
+        g_strfreev (split);
+    }
+
+    apteryx_set_tree (root);
+    apteryx_free_tree (root);
+}
+
+/**
+ * Add an entry to the apteryx database for the specified model
+ *
+ * @param loaded - Structure with the relevant model information
+ */
+void
+yang_library_add_model_information (sch_loaded_model *loaded)
+{
+    GNode *root;
+    GNode *modules;
+    GNode *gnode;
+
+    root = APTERYX_NODE (NULL, g_strdup (YANG_LIBRARY_PATH));
+    modules = add_leaf_strdup (root, YANG_LIBRARY_SCHEMA_MODULE_SET, COMMON_STR);
+    add_leaf_strdup (modules, YANG_LIBRARY_MODULE_SET_NAME, COMMON_STR);
+    if (loaded->model && loaded->model[0] != '\0')
+    {
+        gnode = add_leaf_strdup (modules, YANG_LIBRARY_MODULE_SET_MODULE_PATH,
+                                 loaded->model);
+
+        add_leaf_strdup (gnode, MODULES_STATE_MODULE_NAME, loaded->model);
+        if (loaded->version)
+        {
+            add_leaf_strdup (gnode, MODULES_STATE_MODULE_REVISION,
+                             loaded->version);
+        }
+        if (loaded->ns_href)
+        {
+            add_leaf_strdup (gnode, MODULES_STATE_MODULE_NAMESPACE, loaded->ns_href);
+        }
+        if (loaded->features)
```

Comment:
quite a bit of common code here that could be made more understandable with a small routine